### ISSUE TITLE
Getting kicked out of evil-mode insert for eglot messages

### ISSUE THREAD

**Issue by joncarr** [joaotavora/eglot#221]
When Eglot has a message to share with me in I get kicked out of insert mode.  As an example, I am writing code in Rust and when I type:
```
let tmpl = File::
```
Eglot shares in the minibuffer `[eglot] pub struct File` which I am positive this is the desired behavior.  The only issue is, when that pops up in the minibuffer, I get kicked out of insert mode and put into normal mode.  Is there a setting that I can use to prevent that from happening???

**Comment by joaotavora** [comment]
> mode and put into normal mode. Is there a setting that I can use to prevent that from happening???

I have no idea what is happening, I don't use Evil.  It might have something to do with `eldoc`, the Emacs functionality that Eglot is making use of here.  Maybe ask the Evil developers to tell your where you can put a breakpoint so you can debug the situation?

**Comment by joncarr** [comment]
I'll keep investigating and see if I can find a solution.  My Elisp is horrible but if I can manage to sort it out, Ill be sure to share.  Is Eglot still under active development?  I urge you to keep developing Eglot as it is absolutely wonderful!  I prefer to lsp-mode!!!  Well, i got some digging to do to sort this out.

**Comment by joaotavora** [comment]
> I urge you to keep developing Eglot as it is absolutely wonderful!

It is still under development, but my time is limited :-)

You don't have to know any Elisp to ask the Evil developers if they can suggest something you can trace.  I suppose the Evil developers know the function that quicks you out of "insert mode" and into that other mode.

They can probably teach you how to intrument that function so you can catch the problem in the act (and hypothetically Eglot misbehaving in the backtrace).

**Comment by meqif** [comment]
I also use Evil and cannot replicate the behavior @joncarr is describing. I'd say it's worth attempting to replicate the bug with a fresh Emacs (`emacs -Q`) with just evil-mode and eglot installed.

**Comment by joncarr** [comment]
> I'd say it's worth attempting to replicate the bug with a fresh Emacs (`emacs -Q`) with just evil-mode and eglot installed.

Will give that a go.  I think we can close the issue because I don't think it is eglot related.  I was able to locate the time at which my modes are changed and it is when `eldoc-message-function` is called.  I find it odd however because I receive all kinds of eldoc messages in the message area without any issue.

**Comment by joncarr** [comment]
> I also use Evil and cannot replicate the behavior @joncarr is describing.

Out of curiosity I had a look through your emacs config and noticed you were ignoring a server capability in the eglot config. More specifically it was `(add-to-list 'eglot-ignored-server-capabilites ':hoverProvider)`.

So i copied that line over to my configuration and could no longer replicate the behavior I was describing earlier as well.

**Comment by joncarr** [comment]
Closing, it's my config.  Sorry for the noise.

**Comment by joaotavora** [comment]
@joncarr can you summarize what exactly it was in your config that was causing the problem?  Have you solved it? i.e. have you  gotten it working _with_ the `:hoverProvider` capability?

**Comment by meqif** [comment]
> Out of curiosity I had a look through your emacs config (@meqif ) and noticed you were ignoring a server capability in the eglot config. More specifically it was `(add-to-list 'eglot-ignored-server-capabilites ':hoverProvider)`.

That configuration is not updated, I've reinstated `:hoverProvider` since its output started being truncated to a reasonable size some time ago, and I've been using it since.

**Comment by joncarr** [comment]
> @joncarr can you summarize what exactly it was in your config that was causing the problem? Have you solved it? i.e. have you gotten it working _with_ the `:hoverProvider` capability?

@joaotavora I have not solved it yet, still going through the config, a piece at a time.  I will happily post back once I find the issue responsible!

**Comment by ambihelical** [comment]
@joncarr, did you find out what was causing the problem? 
I am able to reproduce it with the following configuration using ccls and the configuration below. 

```
;;; -*- lexical-binding: t -*-

(require 'package)

(setq package-archives '(("melpa" . "http://melpa.milkbox.net/packages/")
                         ("org"   . "http://orgmode.org/elpa/")
                         ("gnu"   . "http://elpa.gnu.org/packages/")))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))
(unless (package-installed-p 'use-package )
  (package-install 'use-package))
(eval-when-compile
  (require 'use-package))

(use-package eglot 
  :ensure t
  :hook ((c-mode . eglot-ensure)
         (rust-mode . eglot-ensure)
         (c++-mode . eglot-ensure)))

(use-package evil
  :ensure t
  :hook (( prog-mode text-mode ) . evil-mode ))

```
In insert mode within a literal string in a method call within a .cpp file, typing anything causes a message in the message area, which is the comment before the method's declaration in the header file.  After the display, evil reverts to normal mode.

**Comment by joncarr** [comment]
@ambihelical i was not able to discover the exact source of the problem and eventually gave up. I've moved to Emacs LSP and have no issues since. Sorry bud.

**Comment by ambihelical** [comment]
Thanks, no problem.

**Comment by ambihelical** [comment]
@joaotavora here is the connection between eglot, evil and this bug.  It looks like having a hook on text-mode in my config to start evil is the issue (gfm-view-mode derives from text-mode), unless you have another interpretation or some other workaround, this should at least be able to help anyone else with the same problem discovering this closed issue. 

I'm actually not sure how I can enable evil for markdown-mode, since it too is in the mode derivation, but at least this is a workaround. 
```
* evil-normal-state(nil)
  evil-change-state(normal nil)
  evil-change-to-initial-state(nil)
  evil-initialize-state()
  evil-initialize()
  #f(compiled-function (&optional arg) (interactive (list (or current-prefix-arg 'toggle))) #<bytecode 0x3c6289>)(nil)
  ad-Advice-evil-mode(#f(compiled-function (&optional arg) (interactive (list (or current-prefix-arg 'toggle))) #<bytecode 0x3c6289>))
  apply(ad-Advice-evil-mode #f(compiled-function (&optional arg) (interactive (list (or current-prefix-arg 'toggle))) #<bytecode 0x3c6289>) nil)
  evil-mode()
  run-hooks(change-major-mode-after-body-hook text-mode-hook markdown-mode-hook gfm-mode-hook gfm-view-mode-hook)
  apply(run-hooks (change-major-mode-after-body-hook text-mode-hook markdown-mode-hook gfm-mode-hook gfm-view-mode-hook))
  run-mode-hooks(gfm-view-mode-hook)
  gfm-view-mode()
  funcall(gfm-view-mode)
  (progn (funcall mode))
  (condition-case nil (progn (funcall mode)) (error nil))
  (progn (insert string) (condition-case nil (progn (funcall mode)) (error nil)) (font-lock-ensure) (buffer-string))
  (unwind-protect (progn (insert string) (condition-case nil (progn (funcall mode)) (error nil)) (font-lock-ensure) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))
  (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (insert string) (condition-case nil (progn (funcall mode)) (error nil)) (font-lock-ensure) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))
  (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (insert string) (condition-case nil (progn (funcall mode)) (error nil)) (font-lock-ensure) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))))
  (let ((mode x6) (string x4)) (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (insert string) (condition-case nil (progn (funcall mode)) (error nil)) (font-lock-ensure) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))))
  (let* ((val (if (stringp markup) (list (string-trim markup) (intern "gfm-view-mode")) (list (plist-get markup :value) major-mode))) (x4 (car val)) (x5 (cdr val)) (x6 (car x5)) (x7 (cdr x5))) (let ((mode x6) (string x4)) (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (insert string) (condition-case nil (progn (funcall mode)) (error nil)) (font-lock-ensure) (buffer-string)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))))))
  eglot--format-markup("This is called for any underlying socket error")
  (insert ": " (eglot--format-markup documentation))
  (if (string-prefix-p (string-trim documentation) label) nil (goto-char (point-max)) (insert ": " (eglot--format-markup documentation)))
  (progn (setq documentation (match-string 1 documentation)) (if (string-prefix-p (string-trim documentation) label) nil (goto-char (point-max)) (insert ": " (eglot--format-markup documentation))))

```

**Comment by joaotavora** [comment]
> It looks like having a hook on text-mode in my config to start evil is the issue (gfm-view-mode derives from text-mode), unless you have another interpretation or some other workaround, this should at least be able to help anyone else with the same problem discovering this closed issue.

Yep, this is it. This means, if I understand correctly, that if you start `gfm-view-mode` on some buffer, regardless of eglot, the bug will manifest itself, right?  Then it seems like a problem of either `markdown-mode.el` or `evil.el`.

**Comment by ambihelical** [comment]
No not really.  The problem was the way I was ensuring that evil was enabled for certain major modes or derivations of those modes.  I was running evil-mode on prog-mode and text-mode hooks.   It seems that this is a Bad Idea.  Either evil-mode should be run at startup, or evil-local-mode should be run on mode hooks.  I am currently doing the latter, because evil-mode takes a few seconds to start, and I sometimes don't need it right away. 

So I don't think it is a bug in either package. However, it is probably worth a note in eglot's readme.

**Comment by joaotavora** [comment]
> No not really. 

What exactly "not really"? Are you saying (1) the problem doesn't manifest itself when you start `M-x gfm-view-mode` manually in, say, the `*scratch*` buffer. Or (2) it's not a bug at all, just a mis-configuration by your part.

> However, it is probably worth a note in eglot's readme.

This is only sensible of eglot is somehow a necessary step in the reproduction of the problem. If it can be reproduced independently, that note doesn't make much sense.

**Comment by ambihelical** [comment]
(2) is true.  (1) may be true with my misconfiguration, but I haven't tried it. It doesn't happen now that I fixed it of course. 

The problem didn't surface in 3-4 years of almost daily emacs use with the misconfiguration described.  Eglot manifested the issue because it uses another major mode as part of its operation at the same time that I might be in insert mode.  That major mode also has to hook evil-mode via the user's configuration.  This is probably a fairly rare combination since I've never hit it before with any other package. If not for eglot I would have never realized that my evil configuration was not correct. 

IMO,  it is worth a note in the readme (or somewhere linked from the readme), because other evil users may try eglot, but then give up on it because it doesn't seem to be working correctly, when in fact their own configuration is the problem. However I can understand if you think this is out of scope to document.

**Comment by joaotavora** [comment]
> (2) is true. (1) may be true with my misconfiguration, but I haven't tried it. It doesn't happen now that I fixed it of course.

OK. (1) is probably true.

Now you say (2) is true, but I don't agree with you.  I think your configuration is actually reasonably sensible (start a minor mode X whenever this major mode Y kicks in is pretty standard, although I don't know exactly what `use-package` is doing.  I don't use it.).  So I would disagree that it isn't a bug, i.e in my opinion it is a bug, probably in `evil.el` (but I'm not 100% sure, and certainly haven't debugged). But if you find the time to confirm (1), then a bug report to evil is probably worth it.

> IMO, it is worth a note in the readm

It might, yes. I think I'll do something similar, which is mark this (and maybe some more issues) a "known workaround" and link from the README to the github list of such things.

**Comment by ambihelical** [comment]
Ok, I thought this about modes as well until this problem.  Hopefully evil is not an exception to the convention. Perhaps I stopped debugging too early. I'll look into it further if I have time.
